### DISCUSSION TITLE
Has anyone been able to use nuxt 3 with ethers.js?

### DISCUSSION THREAD

**Discussion by jahvi** [nuxt/nuxt#17236] [Questions]
I'm been trying to use use `ethers` in a project but when I try to import it I get the following error:

```
(node:15742) Warning: To load an ES module, set "type": "module" in the package.json or use the .mjs extension.
(Use `node --trace-warnings ...` to show where the warning was created)
/project/node_modules/@ethersproject/signing-key/lib.esm/elliptic.js:1
import BN from 'bn.js';
^^^^^^

SyntaxError: Cannot use import statement outside a module
...
```

By adding the following to the nuxt config file I've been able to get further:

```js
build: {
        transpile: [
            '@ethersproject/signing-key',
            '@ethersproject/basex'
        ],
}
```

But now I have the following in the browser console:

```
elliptic.js?v=87fe3b71:1 Uncaught (in promise) SyntaxError: The requested module '/_nuxt/node_modules/bn.js/lib/bn.js?v=87fe3b71' does not provide an export named 'default'
```

I thought it would be related to vite and the way it bundles dependencies but when using the latest version of vite-cli it works just fine. Even after excluding `bn.js` in the `vite.optimizeDeps` option it still doesn't work.

It looks like nuxt is using vite ^2.8 as well so maybe there's a nuxt specific issue or it could be a problem with these libraries themselves?

**Comment by kazu80** [comment]
I'm using ethers.js to display the eth I have, and I'm not getting any errors.
I did import it like 

```
import {ethers} from "ethers";

...

export const getProvider = (network: ChainName) => {
    ...

    const provider = new ethers.providers.JsonRpcProvider(provider_url)
    return provider
}
```

nuxt.config

```
export default defineNuxtConfig({
    build: {
        transpile: ["@ethersproject", "ethers"],
    },
```

**Reply by jahvi** [reply]
Still getting the same issue with this approach unfortunately

**Reply by mynameisjeffe** [reply]
Same here. Getting this error 
`Uncaught SyntaxError: import not found: default 
[bignumber.js:9:7]
(http://localhost:3000/_nuxt/node_modules/@ethersproject/bignumber/lib.esm/bignumber.js?v=a5cdeeff)`

**Comment by skeithc** [comment]
After hours of trying to get ethers to work at all, I've only been able to make it work by using it only on client-side code through dynamic imports. I tried this because I noticed that the errors I get seem to be related to nuxt trying to import for SSR.

In case it's useful to somebody, the following code works for me in [one of my pages](https://github.com/skeithc/truffle-nuxt-vue-erc721-box/blob/1dd6a109948f04477711d3f23079a0924bdfab9b/pages/index.vue#L4-L27):
```vue
<script setup lang="ts">
let ethers;

async function onMint() {
  if (!ethers) {
    ethers = await import('ethers');
  }

  ...
}
</script>
```

With this change, I didn't need to add ethers and @ethersproject to build.transpile.

What led me to do dynamic imports is because I noticed that it works when I disable ssr in nuxt.config.ts causing me to believe that it's the server side that's having issues.

**Reply by jahvi** [reply]
That would explain why it works with standalone vite since that's all client side, I'll stick to this approach for now thanks for sharing.

**Comment by wobsoriano** [comment]
I’m working on a set of composables for Eth using ethers. It’s a port of wagmi from React.

https://github.com/wobsoriano/vagmi

I’ll try to add a Nuxt example as soon as I finish the docs and test

**Reply by ennioVisco** [reply]
Super nice idea. @wobsoriano let me know if I can help in some way!

**Comment by mynameisjeffe** [comment]
This is a semi-working playground for Nuxt and Ether.js

https://stackblitz.com/edit/github-ar5zza-gs5l5n?file=package.json,nuxt.config.ts,app.vue

Needed to add the following so that Vite will pre-bundle dependencies of bn.js
`
  vite: {
    optimizeDeps: {
      include: ['bn.js', 'js-sha3', 'hash.js', 'aes-js', 'scrypt-js', 'bech32'],
    },
  },
`

I still get this error when I run Nuxt
`WARNING: Missing strong random number source`

But I don't see this error on browser's console.log anymore. So I'm not sure what else is missing.

**Comment by JGJP** [comment]
another alternative is setting `ssr: false` in `nuxt.config.ts`
